amplitude.py

```diff
@@ -1,20 +1,79 @@
+import boto.s3
+import hashlib
+import hmac
 import json
+import os
+import requests
 import sys
 
-def transform(line):
-    try:
-        parsed_line = json.loads(line)
+from threading import Timer
 
-        assert(type(parsed_line) is dict)
+AMPLITUDE_API_KEY = os.environ["FXA_AMPLITUDE_API_KEY"]
+AWS_ACCESS_KEY = os.environ["FXA_AWS_ACCESS_KEY"]
+AWS_SECRET_KEY = os.environ["FXA_AWS_SECRET_KEY"]
+HMAC_KEY = os.environ["FXA_AMPLITUDE_HMAC_KEY"]
 
-        event = parsed_line["Fields"]
+# For crude pre-emptive rate-limit obedience.
+MAX_EVENTS_PER_BATCH = 10
+MAX_BATCHES_PER_SECOND = 100
+BATCH_INTERVAL = 1.0 / MAX_BATCHES_PER_SECOND
 
+def handle (message):
+    # http://docs.aws.amazon.com/AmazonS3/latest/dev/notification-content-structure.html
+    records = json.loads(message)["Records"]
+    for record in records:
+        if record["eventSource"] != "aws:s3":
+            continue
+
+        aws_region = boto.s3.connect_to_region(record["awsRegion"])
+        s3_bucket = aws_region.get_bucket(record["s3"]["bucket"]["name"])
+        s3_object = s3_bucket.get_key(record["s3"]["object"]["key"])
+
+        send(s3_object.get_contents_as_string().splitlines())
+
+def send (events):
+    batch = []
+
+    while len(batch) < MAX_EVENTS_PER_BATCH and len(events) > 0:
+        event = json.loads(events.pop(0))
+
+        # https://amplitude.zendesk.com/hc/en-us/articles/204771828#keys-for-the-event-argument
+        assert("device_id" in event or "user_id" in event)
+        assert("event_type" in event)
         assert("time" in event)
-        assert("user_id" in event or "device_id" in event)
 
-    except:
-        return None
+        insert_id_hmac = hmac.new(HMAC_KEY, digestmod=hashlib.sha256)
+
+        if "user_id" in event:
+            user_id_hmac = hmac.new(HMAC_KEY, event["user_id"], hashlib.sha256)
+            event["user_id"] = user_id_hmac.hexdigest()
+            insert_id_hmac.update(event["user_id"])
+
+        if "device_id" in event:
+            insert_id_hmac.update(event["device_id"])
+
+        if "session_id" in event:
+            insert_id_hmac.update(str(event["session_id"]))
+
+        insert_id_hmac.update(event["event_type"])
+        insert_id_hmac.update(str(event["time"]))
+        event["insert_id"] = insert_id_hmac.hexdigest()
+
+        batch.append(event)
+
+    # https://amplitude.zendesk.com/hc/en-us/articles/204771828#request-format
+    response = requests.post("https://api.amplitude.com/httpapi",
+                             data={"api_key": AMPLITUDE_API_KEY, "event": json.dumps(batch)})
+
+    if response.status_code >= 400:
+        # For want of a better error-handling mechanism,
+        # one failed request fails an entire dump from S3.
+        raise RuntimeError("HTTP {} response")
+
+    if len(events) > 0:
+        # Wait 10 milliseconds before starting the next batch
+        Timer(BATCH_INTERVAL, send, [events]).start()
```

Comment:
If I'm being precise, we're waiting for more than 10 milliseconds here because we don't start the timer until we've received the response and checked it. But there's no need to push right up against the limit, our projected rate falls appreciably inside the threshold.

Reply:
Using `Timer` here will dispatch the function call on a separate thread.  Is that necessary, or can we just `time.sleep(BATCH_INTERVAL)`?

If you want to make efficient use of the difference, you could always do a `startTime = time.time()` at the start of the function, and then `time.sleep(BATCH_INTERVAL - time.time() + startTime)`.  That might help us get through a backfill of events a little faster?  Not sure how much difference it'd make in practice.

Reply:
> Using `Timer` here will dispatch the function call on a separate thread.  Is that necessary, or can we just `time.sleep(BATCH_INTERVAL)`?

Not at all, it just happened to be the first thing I found when I googled `"python setTimeout equivalent"` yesterday. I'll change it to `time.sleep`!

Reply:
> If you want to make efficient use of the difference, you could always do a `startTime = time.time()` at the start of the function, and then `time.sleep(BATCH_INTERVAL - time.time() + startTime)`.  That might help us get through a backfill of events a little faster?  Not sure how much difference it'd make in practice.

I considered that and decided not to bother because it doesn't take into account variations in round-trip time, so natural network variance would mean we violate the 10 ms threshold pretty regularly? I figured it's better to err on the side of caution and then we can come back to it if throughput turns out to be an issue later on.

Reply:
> ...because it doesn't take into account variations in round-trip time...

Although also, I note that the approach here means that one slow request causes all the following batches to queue up in a big traffic jam.

But I don't think we can avoid that without employing multiple threads, and then we'd have all the mess of signalling the response status from each worker thread so that the main thread can `raise`/`return` appropriately when everything is finished. Ewww, threads. :nauseated_face:

Reply:
> I considered that and decided not to bother because it doesn't take into account variations in round-trip time, so natural network variance would mean we violate the 10 ms threshold pretty regularly?

Just to contradict myself, I ended up with a version of this anyway, which does take round-trip time into account. Before sending, I subtract the interval since the last batch (or zero) from `MIN_BATCH_INTERVAL` in my call to `time.sleep`. Seems to work well.

Reply:
> Before sending, I subtract the interval since the last batch (or zero) from `MIN_BATCH_INTERVAL` in my call to `time.sleep`. Seems to work well.

Oh, and I just realised the new timer plays well with multiple records per message, so I'll reinstate the loop that I ditched in 7a4f1ff.